gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -90,17 +115,83 @@ public Fork(TaskContext taskContext, TaskState taskState, Object schema, int bra
     this.writer = buildWriter();
 
+    this.recordQueue = BoundedBlockingRecordQueue.newBuilder()
+        .hasCapacity(taskState.getPropAsInt(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_CAPACITY_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_CAPACITY))
+        .useTimeout(taskState.getPropAsLong(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_TIMEOUT_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_TIMEOUT))
+        .useTimeoutTimeUnit(TimeUnit.valueOf(taskState.getProp(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_TIMEOUT_UNIT_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_TIMEOUT_UNIT)))
+        .collectStats()
+        .build();
+
+    this.countDownLatch = countDownLatch;
+
     this.closer.register(this.rowLevelPolicyChecker);
     this.closer.register(this.writer);
+
+    this.forkState = new AtomicReference<ForkState>(ForkState.PENDING);
   }
 
-  public void processRecord(Object record)
-      throws Exception {
-    for (Object convertedRecord : this.converter.convertRecord(this.convertedSchema, record, this.taskState)) {
-      if (this.rowLevelPolicyChecker.executePolicies(convertedRecord, this.rowLevelPolicyCheckingResult)) {
-        this.writer.write(convertedRecord);
-      }
+  @Override
+  public void run() {
+    setForkState(ForkState.PENDING, ForkState.RUNNING);
+    try {
+      processRecords();
+      setForkState(ForkState.RUNNING, ForkState.SUCCEEDED);
+    } catch (IOException ioe) {
+      setForkState(ForkState.RUNNING, ForkState.FAILED);
+      this.logger.error(
+          String.format("Fork %d of task %s failed to process data records", this.index, this.taskId), ioe);
+      Throwables.propagate(ioe);
+    } catch (DataConversionException dce) {
+      setForkState(ForkState.RUNNING, ForkState.FAILED);
+      this.logger.error(
+          String.format("Fork %d of task %s failed to convert data records", this.index, this.taskId), dce);
+      Throwables.propagate(dce);
+    } catch (Throwable t) {
+      setForkState(ForkState.RUNNING, ForkState.FAILED);
+      this.logger.error(String.format("Fork %d of task %s failed to process data records", this.index, this.taskId), t);
+      Throwables.propagate(t);
+    } finally {
+      // Clear the queue and count down so the parent task knows this fork is done (succeeded or failed)
+      this.recordQueue.clear();
+      this.countDownLatch.countDown();
+    }
+  }
+
+  /**
+   * Put a new record into the record queue for this {@link Fork} to process.
+   *
+   * <p>
+   *   This method is used by the {@link Task} that creates this {@link Fork}.
+   * </p>
+   *
+   * @param record the new record
+   * @return whether the record has been successfully put into the queue
+   * @throws InterruptedException
+   */
+  public boolean putRecord(Object record) throws InterruptedException {
+    if (this.forkState.compareAndSet(ForkState.FAILED, ForkState.FAILED)) {
```

Comment:
The whole "if" is not atomic. What kind of guarantee are we trying to provide here?